### PR TITLE
[pipeline] Added BatchBackend

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

**Comment by jigold** [comment]
@danking Tests are passing!

**Comment by danking** [comment]
Sorry for the delay, this is first PR for me tomorrow.

File: pipeline/pipeline/pylintrc [hunk 1]
```diff
@@ -5,8 +5,9 @@
 # many local variables, W0603 using the global statement, R0902 too many
 # instance attributes, R1705 unnecessary "else" after "return", W0511 fixme,
 # R0903 Too few public methods, R0401 Cyclic import, R0801 Similar lines
+# W0640 cell-var-from-loop
 
-disable=C0111,W1203,W1202,C0111,R0913,W0622,W0212,W0621,R0914,W0603,R0902,R1705,W0511,R0903,R0401,R0801
+disable=C0111,W1203,W1202,C0111,R0913,W0622,W0212,W0621,R0914,W0603,R0902,R1705,W0511,R0903,R0401,R0801,W0640
```
Comment (danking) [inline pipeline/pipeline/pylintrc]: if we make copy_task_inputs an instance method somehow, then we don't have to ignore W0640. I think W0640 is generally a good warning though I understand why you added it. I prefer it's a line-specific ignore though.

File: pipeline/test/test_pipeline.py [hunk 2]
```diff
@@ -74,6 +77,41 @@ def test_single_task(self):
...
+        self.write(input_file, 'abc')
+
+        p = Pipeline()
+        input = p.read_input(input_file)
+        t = p.new_task()
+        t.command(f'cat {input} > {t.ofile}')
+        p.write_output(t.ofile, output_file)
+        p.run()
+
+        self.assertEqual(self.read(output_file), msg)
```
Comment (danking) [inline pipeline/test/test_pipeline.py]: A pytest gives nicer error messages when you just use `assert x == y`

File: pipeline/test/test_pipeline.py [hunk 3]
```diff
@@ -74,6 +77,41 @@ def test_single_task(self):
...
+        self.rm(input_file, output_file)
```
Comment (danking) [inline pipeline/test/test_pipeline.py]: this should be in a finally block, probably? unless you purposely want to leave it when the assert fails

File: pipeline/test/test_pipeline.py [hunk 4]
```diff
@@ -74,6 +77,41 @@ def test_single_task(self):
         self.assertEqual(self.read(output_file), msg)
         self.rm(output_file)
 
+    def test_single_task_w_input(self):
+        msg = 'abc'
+        input_file = '/tmp/data/example1.txt'
+        output_file = '/tmp/test_single_task_w_input.txt'
```
Comment (danking) [inline pipeline/test/test_pipeline.py]: Seems more robust to use:
```
with tempfile.NamedTemporaryFile('w') as input, tempfile.NamedTemporaryFile('rw') as output:
    ...
```

File: pipeline/test/test_pipeline.py [hunk 5]
```diff
@@ -74,6 +77,41 @@ def test_single_task(self):
...
+        self.write(input_file, 'abc')
```
Comment (danking) [inline pipeline/test/test_pipeline.py]: then this can just be `input.write('abc')`

File: pipeline/test/test_pipeline.py [hunk 6]
```diff
@@ -74,6 +77,41 @@ def test_single_task(self):
...
+
+        p = Pipeline()
+        input = p.read_input(input_file)
```
Comment (danking) [inline pipeline/test/test_pipeline.py]: and this would be `p.read_input(input.name)`

File: pipeline/test/test_pipeline.py [hunk 7]
```diff
@@ -159,7 +197,6 @@ def test_scatter_gather(self):
         self.assertEqual(self.read(output_file), '2\n1\n0')
         self.rm(output_file)
 
-
     # FIXME: Run docker inside docker
```
Comment (danking) [inline pipeline/test/test_pipeline.py]: what's the deal with this one?

File: pipeline/test/test_pipeline.py [hunk 8]
```diff
@@ -174,3 +211,75 @@ def test_scatter_gather(self):
...
     #     self.assertEqual(self.read(output_file), msg)
     #     self.rm(output_file)
+
+
+class BatchTests(unittest.TestCase):
+    def pipeline(self):
+        url = os.environ.get('BATCH_URL')
+        if url is None:
+            raise Exception("BATCH_URL not defined in the environment.")
+        return Pipeline(backend=BatchBackend(url),
```
Comment (danking) [inline pipeline/test/test_pipeline.py]: I think `BatchBackend(os.environ['BATCH_URL'])` is fine, this is test code and a KeyError that references this line is pretty clear about what happened.

File: pipeline/test/test_pipeline.py (hunk 7 above)
Comment (jigold) [reply pipeline/test/test_pipeline.py]: You can't do docker run in k8s.

File: pipeline/pipeline/backend.py [hunk 9]
```diff
@@ -90,5 +126,152 @@ def _get_random_name():
...
+                f"  Log:\t{log}\n")
+
+        if failed_jobs or status['jobs']['Complete'] != n_jobs_submitted:
+            raise Exception(fail_msg)
+        else:
+            print("Pipeline completed successfully!")
+
+    def tmp_dir(self):
+        hail_scratch_bucket = 'hail-pipeline-scratch'
+        subdir_name = 'pipeline-{}'.format(get_sha(16).lower())  # FIXME: This should check if directory exists
```
[the remark this answers is not in the document]
Comment (jigold) [reply pipeline/pipeline/backend.py]: We can't fix this until we get the authorization / authentication problem sorted out. Whose service account do we use to check whether a bucket exists or not. I had this before with the gcloud Python library, and you suggested I take it out for this reason.

So my preference would be for it to be an issue. Everything is hard coded now anyways.

File: pipeline/pipeline/backend.py [hunk 10]
```diff
@@ -90,5 +111,151 @@ def _get_random_name():
...
+
+        activate_service_account = 'gcloud auth activate-service-account ' \
+                                   'pipeline-test-0-1--hail-is@hail-vdc.iam.gserviceaccount.com ' \
+                                   '--key-file /secrets/pipeline-test-0-1--hail-is.key'
+
+        for task in pipeline._tasks:
+            def copy_input(r):
+                if isinstance(r, ResourceFile):
+                    if isinstance(r, InputResourceFile):
+                        return [f'gsutil cp {r._input_path} {local_tmpdir}/{r.file_name}']  # FIXME: This is serial!
```
Comment (danking) [inline pipeline/pipeline/backend.py]: I don't think this is a FIXME? If it is let's make an issue instead and assign it to you.

File: pipeline/test/test_pipeline.py (hunk 7 above)
Comment (danking) [reply pipeline/test/test_pipeline.py]: let's delete this and add an issue describing it.

File: pipeline/Makefile [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+	. ../loadconda && conda activate hail-pipeline && python -m pylint --rcfile pipeline/pylintrc pipeline --score=n
+	mkdir -p build && touch $@
+
+test-local: build/conda-env build/batch $(PY_CHECKERS)
+	POD_NAMESPACE='test' BATCH_USE_KUBE_CONFIG=1 ./test-locally.sh
+
+# local means server and test client are two processes on one machine
+# in-cluster means in a k8s pod (from which we get k8s creds)
+test-local-in-cluster: build/conda-env build/batch $(PY_CHECKERS)
+	POD_NAMESPACE='test' ./test-locally.sh
```
[the remark this answers is not in the document]
Comment (jigold) [reply pipeline/Makefile]: I copied it from yours ;)

**Pull request merged**